legion: add missing build dependency and new variants

mpokorny opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
This PR comprises two commits at the moment, addressing distinct functionality. They are small commits, and could be rebased to a single commit after review and before merging the PR.

mpokorny committed:
legion: add pip dependency for build

pip is needed for the build when the Legion Python binding is enabled

mpokorny committed:
legion: add variants for gc logging and system OpenMP use

This also removes the `cmake_cxx_flags` variable from `cmake_args()`
because that variable had no effect.

mpokorny commented:
@spackbot fix style

mpokorny committed:
[@spackbot] updating style on behalf of mpokorny

elliottslaughter commented on var/spack/repos/builtin/packages/legion/package.py:
```diff
@@ -414,9 +419,6 @@ def cmake_args(self):
             options.append("-DLegion_BUILD_BINDINGS=ON")
             options.append("-DLegion_REDOP_COMPLEX=ON")  # required for bindings
 
-        if spec.variants["build_type"].value == "Debug":
-            cmake_cxx_flags.extend(["-DDEBUG_REALM", "-DDEBUG_LEGION", "-ggdb"])
-
```
I agree that the removed code seems odd. I did not put it here and don't know who did. I mainly want to be sure we're not breaking anything. @pmccormick do you know?

mpokorny replied:
AFAICT, the `cmake_cxx_flags` variable exists nowhere in Spack, and it appears to be a local variable, so it cannot have any effect on the build. Nevertheless, it would be good for somebody else to verify it.

pmccormick replied:
Hmmm...  I don't recall it off the top of my head either.  (could be that I've forgotten about it ;-) 

In general, I agree -- seems to have zero impact on the build...

elliottslaughter commented:
@rbberger also usually looks at these.

elliottslaughter commented:
This looks fine to me, mostly just want to confirm one point before we remove the (apparent) old-style debug code.

cmelone commented on var/spack/repos/builtin/packages/legion/package.py:
```diff
@@ -446,6 +448,13 @@ def cmake_args(self):
         # This disables Legion's CMake build system's logic for targeting the native
         # CPU architecture in favor of Spack-provided compiler flags
         options.append("-DBUILD_MARCH:STRING=")
+
+        if "+openmp" in spec and "+sysomp" in spec:
+            options.append("-DLegion_OpenMP_SYSTEM_RUNTIME=ON")
+
+        if "+gc" in spec:
+            options.append("-DCMAKE_CXX_FLAGS=-DLEGION_GC")
+
```
this may require too many changes for this PR, but I believe `in spec` is no longer suggested for use

https://github.com/spack/spack/pull/44348#discussion_r1612192317

mpokorny replied:
I'm happy to replace all the uses of `in spec` with `spec.satisfies(...)`. Should I open another PR just for that, or add another commit to this PR? I don't know what Spack expects for a series of small changes to packages. This single PR could conceivably already be made into three PRs, even without the additional proposed change.

alecbcs replied:
Adding another commit to the existing PR (as you did) is probably my preferred as a reviewer. (So nicely done!)

mpokorny committed:
legion: use spec.satisfies() in cmake_args()

e.g, replace use of '"+foo" in spec' with 'spec.satisfies("+foo")'

mpokorny committed:
legion: avoid configuring with multiple "-DLegion_REDOP_COMPLEX=ON" arguments

In the previous version, when both '+redop_complex' and '+bindings'
was set, two instances of "-DLegion_REDOP_COMPLEX=ON" arguments were
generated for cmake configuration.

mpokorny committed:
legion: fix value of "Legion_OUTPUT_LEVEL" for configuration

the previous version had no effect on setting the configuration value

mpokorny commented:
I'm about to push three more commits to this branch: one to convert syntax to use 'spec.satisfies()' (as suggested by @cmelone), and two more that address additional issues that I identified. Comments welcome on whether I should use more than on PR for these commits, but I'm inclined to use only one (I don't see a lot of packages with sequences of merge commits in the main repo).

elliottslaughter reviewed (approved):
Looks good to me.

mpokorny committed:
Merge branch 'develop' into legion

alecbcs reviewed (approved):
@mpokorny thanks for going in and updating all of the uses of `if in spec` to `spec.satisfies()` this is awesome!
